RangeSlider 구현, 뷰모델 간 데이터 전달방식 수정

junu0516 opened the pull request:
## 💡 작업 내용
- 이전에 논의한대로 클로저가 아닌 `BehaviorRelay` 를 사용하는 방식으로 수정
- `RangleSlider` 구현 후, rx extension을 적용해서 선택한 시간의 시작과 끝 값을 바인딩
- `Double`값을 `TimeComponents`로 변환해서 시,분,초를 각각 계산 후 `CertifiableTime` 값으로 변환해서 다른 뷰모델로 전달

## 💡 고민사항
### 1) 사용자 정의 이벤트(UIControl.Event) 관련
- 현재 구현한 커스텀슬라이더에서, 터치 이벤트를 처리할 때 .valueChanged에 대해서 이벤트를 처리하고 있고, 슬라이더가 내부적으로 가지고 있는 버튼이 2개고, 뷰모델에서도 각 버튼이 가리키는 값에 대응되는 시작과 끝 시간에 해당하는 값을 각각 Subject로 가지고 있습니다.
```swift
sendActions(for: .valueChanged)

//뷰컨트롤러에서의 슬라이더와 뷰모델 간 rx 바인딩
timeRangeSlider.rx.leftValue
    .bind(to: viewModel.selectedStartTime)
    .disposed(by: disposeBag)

timeRangeSlider.rx.rightValue
    .bind(to: viewModel.selectedEndTime)
    .disposed(by: disposeBag)
```
- 문제는 왼쪽 버튼은 그대로 두고, 오른쪽 버튼을 터치했을 때 `rightValue` 뿐만 아니라  `leftValue` 에 대한 .valueChanged 액션도 동시에 발생해서, 기존에 의도한대로 1개의 이벤트만 방출하는 것이 아닌 매번 불필요하게 2개의 이벤트를 항상 방출하는 상황이 발생합니다.
- `.valueChanged` 를 사용하지 않고 `UIControl.Event` 에 leftValueChanged, rightValueChanged.. 이런식으로 커스텀 이벤트를 따로 정의해서, 슬라이더 내부에서 isLeftThumTouched 속성값에 따라 각각 다른 이벤트를 `sendAction(for:)` 하도록 했으나 이럴 경우에는 rx extension에 정의된 코드가 제대로 동작하지 않습니다. 
``` swift
// 커스텀 이벤트 정의
extension UIControl.Event {   
    static var leftValueChanged: UIControl.Event {  UIControl.Event() }
    static var rightValueChanged: UIControl.Event {  UIControl.Event() }
}

extension Reactive where Base: RangeSlider {
    
    var leftValue: ControlProperty<Double> {
        base.rx.controlProperty(
            editingEvents: .leftValueChanged, // 기존에는 .valueChanged로 명시
            getter: { $0.leftValue },
            setter: { _, _ in }
        )
    }
    
    var rightValue: ControlProperty<Double> {
        base.rx.controlProperty(
            editingEvents: .rightValueChanged,  // 기존에는 .valueChanged로 명시
            getter: { $0.rightValue },
            setter: { _, _ in }
        )
    }
}
```
- 우선은 .valueChanged를 그대로 둬서 기능은 제대로 동작하게 했지만, 아무런 변화가 없다면 불필요하게 이벤트를 방출할 필요가 없을 것 같아서 추후 따로 정의한 UIControl.Event 타입 속성이 제대로 동작하도록 하는 법을 찾아서 적용해보려 합니다!
    
### 2) 기존에 선택된 시간 관련 값을, TimeSelectViewController 화면을 띄웠을 때 초기값으로 보여주기
- 기존에 솔이 구현한 ColorPickerViewController과 마찬가지로, 기존에 선택한 값을 기억해뒀다 화면을 띄울 때 초기값으로 보여주려 하는데 잘 적용이 안되네요 ㅋㅋ
- 대강 아래처럼 뷰모델의 생성자로 받은 timeStream 파라미터값을 그대로 바인딩 메소드로 넘겨서 맨 처음에 selectedTime 속성에 바인딩 시키고, 뷰컨트롤러에서는 슬라이더를 바인딩된 뷰모델의 selectedTime 속성값을 읽어들여 다시 바인딩하는 과정을 생각했어요!
```swift
 func bind(
        to navigation: GoalNavigation,
        with timeStream: BehaviorRelay<CertifiableTime>
    ) {
  
        timeStream
            .bind(to: selectedTime)
            .disposed(by: disposeBag)
     }   
```
- 제가 생각하는 문제는 일단 커스텀 슬라이더 내부에서 layoutSubViews()를 아래처럼 오버라이딩하고 있는데 이게 문제인건가 막연하게 생각하고 있습니다.
    - RangeSlider 내부에 leftValue, rightValue 라는 속성을 정의했는데 이것의 초기값이 생성자로 받는 minValue, maxValue 값으로 각각 세팅되있어서 TimeSelectViewModel에서 어떻게 바인딩되는 지에 상관 없이 RangeSlider의 leftValue, rightValue의 초기값은 항상 정해져있기 때문에 고정된 상태로 보이는게 아닐까,, 생각하고 있어요 
```swift
    override func layoutSubviews() {
        super.layoutSubviews()
        // 현재 leftValue, rightValue의 값에 맞춰서 버튼의 위치를 변경
        updateLayout(to: rightValue, direction: .right)
        updateLayout(to: leftValue, direction: .left)
    }
```
- 우선은 이 부분도 레이아웃 그리는 방식과 바인딩 순서를 한 번 다시 정리해보고 어떻게 흐름을 맞출 지 생각해보려 합니다!

## 💡 미리보기
<img src="https://user-images.githubusercontent.com/68586291/188535401-fe1bd6c9-3d12-4914-aa28-a598ad64336d.gif" style="width: 20%;"/>

junu0516 committed:
refactor: [#41] GoalAddViewModel <-> PeriodSelectViewModel 간 데이터 전달 방식 수정

- 특정 뷰모델에서 다른 뷰모델 인스턴스를 생성하는 부분을 모두 제거

junu0516 committed:
refactor: [#41] GoalAddViewModel <-> TimeSelectViewModel 간 데이터 전달 방식 수정

junu0516 committed:
refactor: [#41] 뷰모델 간 데이터 전달 시 BehavirRelay 타입을 전달하도록 다시 수정

junu0516 committed:
feat: [#38] RangeSlider 구현 및 추가

junu0516 committed:
feat: [#38] RangleSlider의 rx 함수 접근을 위한 extension 추가(ControlProperty)

junu0516 committed:
refactor: [#38] RangleSlider의 버튼이 양쪽 끝 위치에 초기화되도록 수정

- layoutSubViews() 호출 시점에 updateLayout 호출

junu0516 committed:
feat: [#38] CertifiableTime, TimeComponent 생성자 추가 정의

junu0516 committed:
feat: [#38] TimeSelectViewController에 RangeSlider 적용 및 TimeSelectViewModel 바인딩 맞춰서 수정

junu0516 committed:
refactor: [#38] 모달 나타내는 방식을 .overFullScreen으로 수정

junu0516 committed:
refactor: [#38] TimeSelectViewModel.selectedTIme 속성 BehaviorRelay로 수정

junu0516 committed:
chore: [#38] 필요한 속성 추가 정의, 오타 수정

Hansolkkim commented on rabit/rabit/Common/Views/RangeSlider.swift:
```diff
@@ -0,0 +1,220 @@
+        isRightThumbTouched = rightThumbButton.frame.contains(touchedPoint)
+        previousTouchedPoint = touchedPoint
+        
+        if isLeftThumbTouched {
+            leftThumbButton.isSelected = true
+        } else {
+            rightThumbButton.isSelected = true
+        }
```
```swift
leftThumbButton.isSelected = isLeftThumbTouched
rightThumbButton.isSelected = isRightThumbTouched
```
이런 식으로 표현할 수 있지 않나 생각이 듭니다!

Hansolkkim replied:
이 부분이 고민 사항 1번에 해당하는 부분인가보네요..ㅎㅎ

junu0516 replied:
넵 맞아요 ㅋㅋ if-else 구문에 지금은 저렇게 값 할당만 있는데 원래는 밑에
```swift
sendActions(for: .leftValueChanged)
```
이런 식으로 메소드를 호출하는 것까지 같이 명시했는데 이벤트가 여기에 맞춰서 발생하지 않더라구요
커스텀 이벤트 만드는 방법이 뭔가 잘못된 것 같은데 찾아봐야겠어요 ㅎㅎ

junu0516 replied:
> ```swift
> leftThumbButton.isSelected = isLeftThumbTouched
> rightThumbButton.isSelected = isRightThumbTouched
> ```
> 
> 이런 식으로 표현할 수 있지 않나 생각이 듭니다!

이 부분은 개선해서 푸시할께요!!

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/GoalAddViewModel.swift:
```diff
@@ -24,10 +22,8 @@ final class GoalAddViewModel: GoalAddViewModelInput, GoalAddViewModelOutput {
-    let selectedPeriod = PublishRelay<Period>()
-    let timeSelectViewModel = PublishRelay<TimeSelectViewModel>()
-    let selectedTime = PublishRelay<CertifiableTime>()
+    let selectedPeriod = BehaviorRelay<Period>(value: Period())
+    let selectedTime = BehaviorRelay<CertifiableTime>(value: CertifiableTime())
```
혹시 얘네를 BehaviorRelay 타입으로 설정하신 이유가 있을까요?

junu0516 replied:
PeriodSelectViewModel, TimeSelectViewModel에서 매칭될 속성의 타입을 `BehaviorRelay`로 사용해서 여기에 맞춰서 설정했어요!
사실 다시 보니깐 GoalAddViewModel 자체만 보면 초기값이 필요한 상황(?)은 아니긴 해서 `PublishRelay`로 둬도 무방하긴 한데.. 혹시 보시기에는 `BehaviorRelay`, `PublishRelay` 중 어느게 나아 보이시나요!?

Hansolkkim replied:
생각해봤는데,
아래에서 말씀하신 것처럼, 한 번 시간을 선택하고 다시 선택하고 싶어서 버튼을 눌렀을 때에 이전에 선택했던 값을 초깃값으로 들고있으면 괜찮겠다는 생각이 들었습니다!
`BehaviorRelay`로 하는게 더 좋을것 같아요!

Hansolkkim commented on rabit/rabit/Goal/GoalAdd/GoalAddViewModel.swift:
```diff
@@ -49,32 +45,15 @@ private extension GoalAddViewModel {
-        periodSelectViewModel
-            .map { $0 }
+            .withUnretained(self)
+            .map { $0.0.selectedPeriod }
```
이 부분 신기하네요.. `$0`은 (viewModel, Void) 타입이고 `$1`은 Void 타입으로 찍히는군요..

junu0516 replied:
저도 당연히 map 하면 파라미터 2개일거라 생각했는데 왜 튜플로 하나만 내려오는지.. 신기하네요 ㅋㅋ

Hansolkkim replied:
ㅋㅋㅋㅋㅋ그러니까요
이게
```swift
map { viewModel, preiod in
    ...
}
```
이런 식으로 사용하면 튜플로 안받기는거 같은데, 입력 값에 대한 정의가 없으면 $0가 튜플이네요 ㅋㅋㅋ

junu0516 replied:
저는 보통은 고차함수 쓸 때 파라미터가 적다 싶으면 따로 파라미터명을 명시하지 않긴 하는데,
혹시 보시기에 $0, $1 이런식으로 사용하는게 가독성면에서 좀 이해하기 힘들다 싶으면 이런 부분은 파라미터명 분명하게 명시하도록 규칙 정해도 좋을듯해요 ㅎㅎ

Hansolkkim replied:
저는 크게 상관없어요ㅋㅋ 한 줄로 표기하는게 더 나은 때도 있고, 파라미터 명을 명시하는게 더 좋을 때도 있다고 생각합니다! 나중에 한 번 얘기해봐요ㅎㅎ

Hansolkkim replied:
새로 알아낸 것들 공유하고자 합니다ㅋㅋㅋ
1. `withUnretained(_:)` 메소드에서 클로저를 받을 수 있습니다.
```swift
selectedPeriod.withUnretained(self) {
    ...
}
```
이런 식으로 `map` 오퍼레이터를 쓰지않아도 `withUnretained(_:)` 메소드에서 클로저를 받기 떄문에, 반환할 걸 결정해줄 수 있더라구요!

2. `map(_:)` 오퍼레이터에서 받는 클로저에서 $0, $1과 같이, 달러 표기로 사용하는 파라미터의 개수에 따라서 타입이 바뀌더라구요.
    - `$0`만 사용할 때
        ![SS2022-09-06PM09 10 16](https://user-images.githubusercontent.com/92504186/188631672-2cf6b73d-be74-487a-a58c-f924e7c3b39d.jpg)
    - `$0, $1` 모두 사용할 때
        ![SS2022-09-06PM09 10 48](https://user-images.githubusercontent.com/92504186/188631768-b58dc4aa-839c-4c3e-a37f-1f41654a8603.jpg)

 위 사진처럼, $0만 사용하려고 할때는 $0이 튜플 타입이고, $0, $1을 모두 사용하려고 하면 $0은 viewModel타입이더라구요!

Hansolkkim commented on rabit/rabit/Goal/GoalCoordinator.swift:
```diff
@@ -88,44 +88,49 @@ final class GoalCoordinator: Coordinator, GoalNavigation {
+        let viewModel = PeriodSelectViewModel(navigation: self, with: periodStream)
         let viewController = PeriodSelectViewController(viewModel: viewModel)
-        viewController.modalPresentationStyle = .overCurrentContext
+        viewController.modalPresentationStyle = .overFullScreen
```
```swift
viewController.modalPresentationStyle = .overFullScreen
```
이 부분은 혹시 어떤 이유로 바꾸신건가요?!

junu0516 replied:
아 ㅋㅋ 이게 그 BottomSheet에 스와이핑 비슷한 제스쳐를 취했을 때, 전체 모달도 같이 약간씩 딸려서 내려가는 부분을 없애려고 바꾼 부분입니다!

기존에 .overCurrentContext 로 지정하면, dimmedView와 같이 PeriodSelectViewController 화면에 모달처럼 올라오는데, 이걸 .overFullScreen으로 하면 전체 화면을 아얘 덮어버리기 때문에 애초에 스와이핑이 일어나도 전체 화면도 딸려 내려가는 현상은 발생하지 않더라구요 ㅎㅎ
(dimmedView가 어느정도 투명하게 설정되있기 때문에 .overFullScreen으로 해도 육안으로 봤을 때 이전의 화면 형태와 달라진 점은 없어보였습니다!)

Hansolkkim replied:
사실 아까 말씀하셨던 부분 정리해놓으면 좋을 것 같아서 재질문했습니다ㅋㅋㅋ

Hansolkkim commented on rabit/rabit/Goal/TimeSelect/TimeSelectViewController.swift:
```diff
@@ -67,23 +72,27 @@ final class TimeSelectViewController: UIViewController {
+            .when(.ended)
             .withUnretained(self)
-            .bind { viewController, _ in
+            .bind { viewController, temp in
```
temp를 와일드카드 처리하셨다가 다시 살리신 이유가 있을까요?

junu0516 replied:
ㅋㅋ 이전에 print 찍으려고 처리한 부분인데 지우는걸 깜빡했네요.. 이것도 고쳐서 커밋 푸시할께요 ㅎㅎ

Hansolkkim commented on rabit/rabit/Goal/TimeSelect/TimeSelectViewModel.swift:
```diff
@@ -6,45 +6,57 @@ protocol TimeSelectViewModelInput {
-                    .share()
+        Observable
+            .combineLatest(
+                selectedStartTime.asObservable(),
+                selectedEndTime.asObservable()
```
`asObservable()`을 추가했을 때랑 안했을 때는 어떤 차이가 있나요?

junu0516 replied:
이건 아얘 생각못했는데, 추가 안했을 때도 확인해보고 둘이 뭐가 다른지 학습해봐야겠어요!

Hansolkkim replied:
제가 봤을 땐 둘이 다른게 없는것 같았어요! 어짜피 두 프로퍼티 모두 Observable의 하위 타입인 PublishRelay/ BehaviorRelay여서, 달라지는게 없는 것 같아서 여쭤봤습니다ㅎㅎ

junu0516 replied:
아하 ㅋㅋ 게더에서 알려주신대로 Observable이면서 Observer이니까 굳이 Observable로 불필요하게 변환할 필요가 없군요!
알려주셔서 감사합니다 ㅋㅋ 고쳐서 다시 푸시했어요 커밋

Hansolkkim commented on rabit/rabit/Goal/TimeSelect/TimeSelectViewModel.swift:
```diff
@@ -6,45 +6,57 @@ protocol TimeSelectViewModelInput {
+        with timeStream: BehaviorRelay<CertifiableTime>
+    ) {
+  
+        timeStream
+            .bind(to: selectedTime)
+            .disposed(by: disposeBag)
```
이 부분이 필요한가 생각이 듭니다!

timeStream의 변경 -> selectedTime의 변경

해당 바인딩 코드를 추가하신 이유가 있을까요?

Hansolkkim replied:
고민 사항 2번에 이 부분이 해당하는것 같은데,
제 생각에는 `TimeSelectViewController`에서 selectedTime(선택된 Time Range 관련 프로퍼티)와 left, right Thumb (Slider의 양쪽 Thumb) 간의 바인딩이 존재하지 않아서 바로 표시가 안되는거 아닐까? 생각이 들었어요!

junu0516 replied:
언급하신대로 지금 상황에서는 필요없는 바인딩인 것 같아요

그 고민 2번 관련해서 TimeSelectViewController의 timeRangeSlider 속성의 초기값 설정을 위해 뒀던 코드인데, 의도한대로 동작하지 않아서 우선은 제거하겠습니다 ㅋㅋ

junu0516 replied:
> 고민 사항 2번에 이 부분이 해당하는것 같은데, 제 생각에는 `TimeSelectViewController`에서 selectedTime(선택된 Time Range 관련 프로퍼티)와 left, right Thumb (Slider의 양쪽 Thumb) 간의 바인딩이 존재하지 않아서 바로 표시가 안되는거 아닐까? 생각이 들었어요!

아아 사실 원래 코드는 언급해주신 left, right에 대한 바인딩도 같이 정의해놨었는데 이것도 동작을 안하더라구요 ㅜㅠㅠ 오후에 저희 다시 만났을 때 시간되시면 아얘 상황을 공유드릴께요!

Hansolkkim replied:
그렇군요ㅠㅠ넵넵 알겠습니다!
조금이따 뵐게요ㅎㅎ

Hansolkkim reviewed (commented):
@junu0516 

고생하셨습니다!
리뷰라기 보다는 궁금한 부분 질문드렸습니다ㅎㅎ

junu0516 commented:
> @junu0516
> 
> 고생하셨습니다! 리뷰라기 보다는 궁금한 부분 질문드렸습니다ㅎㅎ

리뷰 지금 읽었습니다! 점심먹고 코멘트달고 수정할 커밋 올릴께요 ㅎㅎ
ㅋㅋㅋ 게더에서 저희 헤어지자마자 기록용으로 PR 다시 작성했는데, 고민에 적은거랑 리뷰에서 언급해주신거랑 겹치네요 ㅎㅎ
정확한 트러블슈팅이군요 ㅋㅋㅋ 🤩

junu0516 committed:
refactor: [#38] ThumbButton의 isSelected 속성값 정하는 로직 수정

junu0516 committed:
chore: [#38] 불필요한 코드 제거

- TimeSelectViewModel의 불필요한 바인딩 우선 제거, asObservable() 불필요하게 쓰인부분 제거
- TimeSelectViewController의 rx바인딩에서 불필요한 파라미터 선언 와일드카드로 대체

junu0516 commented:
@Hansolkkim 
우선 개선사항이랑 불필요한 코드 피드백 주신 부분 수정해서 커밋 푸시했어요!
고민사항 2번에 해당하는 부분은 우선 후순위로 미뤄놓고 연휴동안 좀 더 고민해보려 합니다 ㅎㅎ(추석 끝나고 오프라인으로 만날 때 정확히 상황 공유할께요!)

편하실 때 확인하시고 혹시 추가적으로 피드백 있거나 수정 필요한 부분 있음 말씀해주세요!
